src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/StoredProcedures/HardDelete/hardDeleteIdentityProvider.js

```diff
@@ -0,0 +1,139 @@
+        RequestEntityTooLarge: `${ErrorCodes.RequestEntityTooLarge}: The request could not be completed.`,
+        NotFound: `${ErrorCodes.NotFound}: The id is not found.`,
+        LastIdentity: `${ErrorCodes.BadRequest}: Cannot delete as only one identity provider exists.`,
+        PreconditionFailed: `${ErrorCodes.PreconditionFailed}: One of the specified pre-condition is not met.`
+    };
+
+    // Validate input
+    if (!id) {
+        throw new Error(errorMessages.IdNull);
+    }
+
+    let deletedResourceIdList = new Array();
+
+    getIdentityProviderCountAndDelete(eTag);
+
+    function getIdentityProviderCountAndDelete(eTag) {
+
+        var doc;
+        // Find the resource and all of its history.
+        let query = {
+            query: "SELECT COUNT(r.id) docCount FROM ROOT r"
+        };
+
+        let isQueryAccepted = collection.queryDocuments(
+            collection.getSelfLink(),
+            query,
+            {},
+            function (err, documents, responseOptions) {
+                if (err) {
+                    throw err;
+                }
+
+                if (!documents || !documents.length) {
+                    return;
+                } else {
+                    doc = documents[0];
+                    if (doc.docCount === 0) {
+                        throw new Error(errorMessages.NotFound);
+                    }
+                    else if (doc.docCount === 1) {
+                        throw new Error(errorMessages.LastIdentity);
+                    }
+                    tryQueryAndHardDelete(eTag);
+                }
+            });
+
+        if (!isQueryAccepted) {
+            // We ran out of time.
+            throw new Error(errorMessage.requestEntityTooLarge);
+        }
+
+    }
+
+    function tryQueryAndHardDelete(eTag) {
+        // Find the resource and all of its history.
+        let query = {
+            query: "SELECT * FROM ROOT r WHERE r.id = @id",
+            parameters: [{ name: "@id", value: id }]
+        };
+
+        let isQueryAccepted = collection.queryDocuments(
+            collection.getSelfLink(),
+            query,
+            {},
+            function (err, documents, responseOptions) {
+                if (err) {
+                    throw err;
+                }
+
+                if (documents.length > 0) {
+                    // Delete the documents.
+                    tryHardDelete(documents, eTag);
+                } else {
+                    // There is no more documents so we are finished.
+                    response.setBody(deletedResourceIdList);
+                }
+            });
+
+        if (!isQueryAccepted) {
+            // We ran out of time.
+            throw new Error(errorMessage.requestEntityTooLarge);
+        }
+    }
+
+    function tryHardDelete(documents, eTag) {
+        if (documents.length > 0) {
+            deletedResourceIdList.push(documents[0].id);
```

Comment:
The query above query by id and therefore we will have either 0 or 1 document. I don't think we need all of this logic to push into array and return?

In fact, since it's just deleting one document, you don't really need stored procedure? Is there additional things that will be deleted later or something?

Reply:
The idea was to put additional validation on delete within the stored procedure. There are certain cases that we want to protect against such as not deleting the last identity provider.

Reply:
Updated the stored proc.